### DISCUSSION TITLE
RTSP + USB Audio - Simultaneous Sources?

### DISCUSSION THREAD

**Discussion by domenicdistefano** [mcguirepr89/BirdNET-Pi#1006] [Q&A]
Based on the text from the RTSP Stream box in the Advanced Settings page I don't think this is possible but thought I would ask anyway...

I have a second microphone in a different area of my yard I want to "merge" into the system.  My idea is to place a second Pi with only a microphone in the secondary location and (TBD) figure out how to output it via an RTSP stream. If I could then ADD that source to BirdNET I would be in business. 

Unfortunately again the text below seems clear that its an either or situation **but** it does say multiple RTSP streams can be added simultaneously so perhaps a couple of cheap Pi zeros as "audio servers" is the answer and simply not using the USB sound card on the BirdNET at all?

------------------------
_If you place an RTSP stream URL here, BirdNET-Pi will use that as its audio source.
Multiple streams are allowed but may have a impact on rPi performance.
Analyze ffmpeg CPU/Memory usage with top or htop if necessary.
To remove all and use the soundcard again, just delete the RTSP entries and click Save at the bottom._

**Comment by srd424** [comment]
I think you're right, the way the code is currently set-up I think it's either-or. Though in theory it shouldn't be too much re-architecting to allow both.

At the moment I use Pi Zero to push audio to my main general-purpose house server with the following gstreamer pipeline:
```
gst-launch-1.0 alsasrc provide-clock=false device=default:CARD=Microphone ! opusenc bitrate=48000 ! rtspclientsink location=rtsp://server.lan:8554/mic1
```
I chose gstreamer just because it seemed to pull in far fewer dependencies than ffmpeg, but it shouldn't be hard to do with ffmpeg either. I use [mediamtx](https://github.com/bluenviron/mediamtx) to push the RTSP to, though in theory the icecast instance that BNPI already creates could be used as well I think (although maybe not with RTSP.)

If you did want to mix local and remote audio, the quickest hack would probably be to present even the local audio as an RTSP/HTTP stream to birdnet. Probably adds a little overhead moving the data around unnecessarily, but hopefully not much.

**Comment by domenicdistefano** [comment]
This is a brilliant idea! I can just create RTSP Steam 1 from my existing BirdNET Pi and loop it back locally and then throw in a Pi zero with a second mic as RTSP 2.

I am a little confused however in terms as how you are implementing it. So above you have you gstreamer command line you are using. So how does mediamtx fit in? Can you reply with your full recipe?

Thanks!

**Reply by srd424** [reply]
You can grab mediamtx binaries from its github page - looks like there are multiple arm versions, so you'll need to pick the right one for your model of Pi. There's a default config file in the tarball, I haven't needed to change that, but I guess you might in theory want to set a password if there are untrusted devices/users on your network.

**Reply by srd424** [reply]
If you're "looping back" audio from the Pi the BirdNET-Pi is installed on, you could probably use ffmpeg instead of gstreamer. I'm about to sit down to supper, but I'll see if I can concoct an appropriate ffmpeg command line today or tomorrow :)

**Comment by domenicdistefano** [comment]
Manny thanks & take your time. So gstreamer (or ffmpeg) creates the audio then mediamtx takes that feed and creates the RTSP stream that would be pulled from the BirdNET pi? In my case my BirdNET pi is a RP 4 (where I am plaining on looping back the local audio as you suggested) and I found an older RP 3 I am going to connect the new mic thus producing the second RTSP feed to BirdNet.

**Reply by srd424** [reply]
Something like this might work if you wanted to use ffmpeg instead of gstreamer:
```
ffmpeg -nostdin -f alsa -i default -acodec libmp3lame -b:a 160k -ac 2 -content_type 'audio/mpeg' -f rtsp rtsp://server.lan:8554/starlite
```
Untested on an actual Pi, but it seems to work on my laptop :)

**Comment by domenicdistefano** [comment]
So does the above ffmpeg command both encode the audio from the local USB sound card mic and produce the RTSP stream itself? I am confused how you are mediamtx fits in.

I got my old RP 3 going but jut need the microphone to arrive from Micbooster. I can probably test however in the meantime somewhat.

**Comment by domenicdistefano** [comment]
I think I have all of this figured out! Just to recap my goal is to utilize the local USB microphone installed on my BirdNET Pi (lets call it RP1) **PLUS** add a second remote USB microphone located on a secondary Pi, not running BirdNET (lets call it RP2).  Net result is to have multiple microphones simultaneously feeding the single instance of BirdNET audio so detections can be made originating from either microphone which are located at separate locations. In a perfect world I would want to have multiple independent "nodes" or instances of BirdNET running at each location somehow consolidated into a "master" BirdNET Pi. This however does seem to work but I just got it working and I need to let it run for a few days in production to make sure the wheels don't fall off or at least the steering wheel doesn't shake too much.

So here is my final (pending tweaks) recipe. Many thanks to srd424 for his suggestions above.

**Secondary USB microphone connected to secondary Pi (RP2)**
Note - this Pi does NOT have BirdNET running and only has the minimum software installed in order to produce an RTSP feed.

**Step 1** - Install the Raspberry Pi OS Lite (64-bit) as per the official BirdNET instructions but just do not install BirNET. Pro Tip - For reliability & speed purposes install the OS on an external SATA drive using a SATA to USB 3 adapter. A 512GB drive is only about $25-$30 on Amazon these days with the adapter cable running $8.

**Step 2** - (still operating on RP2) Install ffmpeg & mediamtx

- First login as the regular user, update the OS as usual, set the root password, enable SSH login by Root, and then re-login as root.
- Next install ffmpeg `apt-get install ffmpeg`
- Install lsof `apt-get install lsof`
- Install mediamtx making sure you are in the /root directory 
`wget -c https://github.com/bluenviron/mediamtx/releases/download/v1.0.0/mediamtx_v1.0.0_linux_arm64v8.tar.gz -O - | sudo tar -xz`
- Test to see if you can create a RTSP feed with audio streaming from your USB microphone. In my case this is the particular code that worked: 
```
./mediamtx &
ffmpeg -nostdin -f alsa -ac 1 -i hw:1 -acodec libmp3lame -b:a 160k -ac 2 -content_type 'audio/mpeg' -f rtsp rtsp://localhost:8554/birdmic -rtsp_transport tcp
```
- To see if your RTSP feed is working fire up VLC, go to Media, Open Network Stream, and paste in `rtsp://hostnameofRP2:8554/birdmic` and hit enter. You should connect and be able to hear your microphone.
- Still in the /root directory create a startmic.sh file `nano startmic.sh`
- The contents of the file is the same code used above to test.
- Make the file executable `chmod +x startmic.sh` 
- Execute the crontab command `crontab -e` and select nano as your editor.
- Paste in `@reboot /home/linuxconfig/startmic.sh` then save and exit nano.
- Reboot the Pi and test again with VLC to make sure the RTSP stream is live.

**Step 3** - Reconfigure your BirdNET Pi (RP1)

- Go into your BirdNET Tools, Settings, Advanced Settings and add two entries under the RTSP section. rtsp://localhost:8554/birdmic & rtsp://ipaddressofRP2:8554/birdmic
- Reboot your BirdNET Pi
- Again login as the regular user, update the OS as usual, set the root password, enable SSH login by Root, and then re-login as root.
- Install ffmpeg `apt-get install ffmpeg`
-  Install lsof `apt-get install lsof`
- Install mediamtx making sure you are in the /root directory 
`wget -c https://github.com/bluenviron/mediamtx/releases/download/v1.0.0/mediamtx_v1.0.0_linux_arm64v8.tar.gz -O - | sudo tar -xz`

- You now need to adjust a port used by mediamtx on your BirdNET Pi (RP1) because it conflicts with the same port already in use.
Fire up Nano and edit the mediamtx.yml `nano /root/mediamtx.yml` and search for the line that reads hlsAddress: :8888 and change 8888 to 7777 then save and exit.
- Test to see if you can create a RTSP feed with audio streaming from your USB microphone. In my case this is the particular code that worked: 
```
./mediamtx &
ffmpeg -nostdin -f alsa -ac 1 -i hw:1 -acodec libmp3lame -b:a 160k -ac 2 -content_type 'audio/mpeg' -f rtsp rtsp://localhost:8554/birdmic -rtsp_transport tcp
```
- To see if your RTSP feed is working fire up VLC, go to Media, Open Network Stream, and paste in `rtsp://hostnameofRP1:8554/birdmic` and hit enter. You should connect and be able to hear your microphone.
- Still in the /root directory create a startmic.sh file `nano startmic.sh`
- The contents of the file is the same code used above to test.
- Make the file executable `chmod +x startmic.sh` 
- Execute the crontab command `crontab -e` and select nano as your editor.
- Paste in `@reboot /home/linuxconfig/startmic.sh` then save and exit nano.
- Reboot the Pi and test again with VLC to make sure the RTSP stream is live.

Everything should now be working! To verify hit the Spectrogram on the main BirdNET menu select each each of your RTSP feeds which should be showing in a drop down menu. As long as you hear audio from each you are in business.

My secondary microphone is a couple of weeks out shipping from the UK but in the meantime I am going to monitor performance and make sure the microphone on RP1 performs just as well as when it was "natively"  connected as now its looped back via RTSP.

**Reply by domenicdistefano** [reply]
**Update** - sometimes due to the nature of USB  the `hw:` parameter can "float" randomly between 1&4 upon a reboot Fix is just to repeat the command four times as one will eventually be the one you want.

```
./mediamtx &
ffmpeg -nostdin -f alsa -ac 1 -i hw:4 -acodec libmp3lame -b:a 160k -ac 2 -content_type 'audio/mpeg' -f rtsp rtsp://localhost:8554/birdmic -rtsp_transport tcp
ffmpeg -nostdin -f alsa -ac 1 -i hw:3 -acodec libmp3lame -b:a 160k -ac 2 -content_type 'audio/mpeg' -f rtsp rtsp://localhost:8554/birdmic -rtsp_transport tcp
ffmpeg -nostdin -f alsa -ac 1 -i hw:2 -acodec libmp3lame -b:a 160k -ac 2 -content_type 'audio/mpeg' -f rtsp rtsp://localhost:8554/birdmic -rtsp_transport tcp
ffmpeg -nostdin -f alsa -ac 1 -i hw:1 -acodec libmp3lame -b:a 160k -ac 2 -content_type 'audio/mpeg' -f rtsp rtsp://localhost:8554/birdmic -rtsp_transport tcp

```

**Reply by bagrcoufal** [reply]
Thanks for good idea with microphone stream. But I have problem with script. Only command working perfectly: 
`./mediamtx & ffmpeg -nostdin -f alsa -i default -acodec libmp3lame -b:a 160k -ac 2 -content_type 'audio/mpeg' -f rtsp rtsp://localhost:8554/birdnet `
The same command in execute in script. not working.
```
#!/bin/bash
/home/pi/BirdNET-Pi/mediamtx & /usr/bin/ffmpeg -nostdin -f alsa -i default -acodec libmp3lame -b:a 160k -ac 2 -content_type 'audio/mpeg' -f rtsp rtsp://localhost:8554/birdnet
```
error [alsa @ 0x55ab3f8630] cannot open audio device default (No such file or directory)
default: Input/output error
Any Idea? Thanks

**Reply by iamjoshk** [reply]
Almost 2 years old, but this comment was exactly what I needed to help me get started on setting up an RPi0 2w with a USB mic as a second audio source.

**Comment by Peterkal2112** [comment]
> **Update** - sometimes due to the nature of USB the `hw:` parameter can "float" randomly between 1&4 upon a reboot Fix is just to repeat the command four times as one will eventually be the one you want.
> 
> ```
> ./mediamtx &
> ffmpeg -nostdin -f alsa -ac 1 -i hw:4 -acodec libmp3lame -b:a 160k -ac 2 -content_type 'audio/mpeg' -f rtsp rtsp://localhost:8554/birdmic -rtsp_transport tcp
> ffmpeg -nostdin -f alsa -ac 1 -i hw:3 -acodec libmp3lame -b:a 160k -ac 2 -content_type 'audio/mpeg' -f rtsp rtsp://localhost:8554/birdmic -rtsp_transport tcp
> ffmpeg -nostdin -f alsa -ac 1 -i hw:2 -acodec libmp3lame -b:a 160k -ac 2 -content_type 'audio/mpeg' -f rtsp rtsp://localhost:8554/birdmic -rtsp_transport tcp
> ffmpeg -nostdin -f alsa -ac 1 -i hw:1 -acodec libmp3lame -b:a 160k -ac 2 -content_type 'audio/mpeg' -f rtsp rtsp://localhost:8554/birdmic -rtsp_transport tcp
> ```
Brilliant...

Why not use "default" as Hardware... if the HW parameter float randomly, the default can be a solution...

do you have these steps also for docker ?

I do not want to install ffmpeg and mediamtx, easier and faster is to set up a docker compose file and fire them up....

I like BirdNet-Pi but it is a shame that it does not have a docker image, or the install script supports only Debian 11 and not 12...

**Comment by domenicdistefano** [comment]
Not sure if I ever tried "default"... Also when I was working with this that was before my "Docker awakening"...

I would LOVE a docker version of all of this.

**Reply by srd424** [reply]
My dockerized version mostly works, it just needs documenting properly! Unfortunately I got overtaken by health issues and other circumstances so everything rather got paused. In particularly I think I haven't documented the "multi site" support stuff, so what docs there are are a little out of date :(

My pi zero which does the sound recording is now actually running Balena OS, so I in theory I should be able to publish the sound recording bit as an app on Balena Hub or something, too.

**Reply by srd424** [reply]
FWIW on the docker front, if you haven't seen it, someone else seems to be working on this too: https://github.com/mcguirepr89/BirdNET-Pi/issues/211#issuecomment-2081183546

Meanwhile I'm still distracted by the usual pile of real-world stuff, sorry :(

**Reply by alexbelgium** [reply]
Hi, my docker version works but is heavily dependent on homeassistant - it could however easily be ported in a more general manner by adapting the dockerfile :) for example, mapping of folders, permissions... Is managed by HA so not included out of the box in my version. But if we add it in my dockerfile then it would achieve both aims with a single container :) or don't hesitate to take part of my code if useful to create another standalone container!

**Comment by alexbelgium** [comment]
That's really great! Is there an impact on sound quality to perform this additional encoding to rtsp ? thanks !

**Reply by srd424** [reply]
RTSP itself is just the transport, but whatever codec you choose to stream will lose a bit of quality (unless you use flac or plain wav.) I have a vague memory that when I originally set this up with gstreamer & opus as above, I may have tested to make sure I didn't see a significant drop in detections over the same length of time in similar weather conditions etc. At some point I intend to move to mp3 because it seems to consume significantly less cpu on a pi zero, I'll try to remember to do the same test again.

**Comment by lolorc** [comment]
that looks interesting, I'm also playing with mediamtx & go2rtc, I've just replaced pulseaudio/arecord/ffmpeg(livestream) by one ffmpeg instance doing the wav recordings directly from alsa and also the opus audio encoding to mediamtx. That makes it more efficient to stream video and audio to frigate for instance.

**Reply by srd424** [reply]
Ah, nice, that's something I'd been meaning to look at! I also think saved recordings get converted back to mp3 from the wav? So there might be another step there that could be saved ..

**Reply by lolorc** [reply]
converted to opus to be streamed, saved in wav, analyzed in wav, then converted to mp3.
could stream and save in opus, but that would require decoding it to analyze
for an external source it's decoded in wav, analyzed then encoded again, isn't it ?

**Comment by alexbelgium** [comment]
Thanks for the guide! I've adapted the code to use pcm : 
`ffmpeg -nostdin -f alsa -acodec pcm_s16le -ac 2 -ar 48000 -i hw:1,0 -f rtsp -acodec pcm_s16be rtsp://localhost:8554/birdmic -rtsp_transport tcp`

**Comment by dumbledorethor** [comment]
Thanks for the guide! Here is the config I used for RPI Zero W2 with BOYA BY-LM40 microphone.


./mediamtx &
ffmpeg -nostdin -f alsa -acodec pcm_s24le -channels 1 -ar 48000 -i hw:1,0 -f rtsp rtsp://localhost:8554/birdnet -rtsp_transport tcp

**Reply by kongjudas** [reply]
Thanks! That worked for my BY-LM40 but with a huge delay. Do you know how to get it to work with libopus? I get an error no matter what I try. This is my latest attempt (**Don't use it, not working**):

```
./mediamtx &
ffmpeg -nostdin -f alsa -ac 1 -ar 48000 -i hw:1,0 -c:a libopus -b:a 32k -f rtsp rtsp://localhost:8554/birdnet -rtsp_transport tcp
```

**Comment by Bongo2** [comment]
Thanks for this information.

I use Birdnet-Pi in the docker. I set up AudioMoth as a USB Miktofon. For me this code works very well:
`ffmpeg -nostdin -f alsa -i default -acodec libmp3lame -b:a 160k -ac 2 -content_type 'audio/mpeg' -f rtsp rtsp://localhost:8554/birdnet 
`
Finally the audio in the GUI is played loudly and clearly. Without that I also have to put my speakers to 100%. Before that, I used the RTSP stream of my surveillance camera.

Recently there is also BattyBirdnet-PI in the docker from @alexbelgium. Of course, I would like to use AudioMoth with a sample rate of 384kHz.
If I understand it correctly, the above code only arrives 48kHz. That doesn't work for bats.

Can someone say how I can use the full sample rate?

**Reply by alexbelgium** [reply]
Hi, for bats I'm using `ffmpeg -nostdin -f alsa -acodec pcm_s16le -ac 2 -ar 256000 -i hw:1,0 -f rtsp -acodec pcm_s16be rtsp://localhost:8554/birdmic -rtsp_transport tcp` : pcm keeps higher frequencies than the lossy format mp3. The 256khz should be enough but you could use 384000 if needed. Be sure to adapt the sample rate in battybirdnet-pi and use wav as extraction format if you want to be able to reuse the extracted files : by default mp3 extraction format (post analysis) seems to limit frequencies around 22khz

**Reply by Bongo2** [reply]
My Raspberry-Pi 3b had only USB 2.0.
I am now switching to Raspberry 4B with USB 3.0.
The hardware device is listed with `arecord -L`:

```
null
    Discard all samples (playback) or generate zero samples (capture)
hw:CARD=Microphone,DEV=0
    384kHz AudioMoth USB Microphone, USB Audio
    Direct hardware device without any conversions
plughw:CARD=Microphone,DEV=0
    384kHz AudioMoth USB Microphone, USB Audio
    Hardware device with all software conversions
default:CARD=Microphone
    384kHz AudioMoth USB Microphone, USB Audio
    Default Audio Device
sysdefault:CARD=Microphone
    384kHz AudioMoth USB Microphone, USB Audio
    Default Audio Device
front:CARD=Microphone,DEV=0
    384kHz AudioMoth USB Microphone, USB Audio
    Front output / input
dsnoop:CARD=Microphone,DEV=0
    384kHz AudioMoth USB Microphone, USB Audio
    Direct sample snooping device
```
For me, the input device works `-i default:CARD=Microphone`.

@alexbelgium 
When I use your stream data:
`ffmpeg -nostdin -f alsa -i default:CARD=Microphone -acodec pcm_s16le -ac 2 -ar 256000 -f rtsp -acodec pcm_s16be -rtsp_transport tcp rtsp://localhost:8554/birdmic`

Comes this output:

![hc_3308](https://github.com/user-attachments/assets/b1b72925-388f-4bbe-82cc-e5a9c33822ae)

The output stream is 256000 Hz. That is fine.
The input stream is 48000 Hz. Actually, it should be 384000 Hz. That's how I set it up.
Does it also come by you 48000 Hz? Does that still work?

**Reply by alexbelgium** [reply]
Then perhaps `ffmpeg -nostdin -f alsa -ar 384000 -ac 2 -i plughw:1,0 -acodec pcm_s16le -f rtsp -rtsp_transport tcp -ar 256000 -acodec pcm_s16be rtsp://localhost:8554/birdmic` ? plughw should adapt the input to make sure it matches what you request

**Reply by Bongo2** [reply]
Thanks for your help.
I have now set my Audiomoth to 250 kHz.
-ar 250000 must be in the input section. Then it works.

For me it works with:
```
-i plughw:CARD=Microphone
and
-i default:CARD=Microphone
```
If birdnet or battybirdnnetnet is installed, it only works with:
`-i default
`
I think `arecord -L` can be very helpful.

My plan was Birdnet-Pi and Battybirdnet Pi to run in the docker. But BattyBirdnet-PI sometimes makes CPU utilization of 99%. And the data from the previous day were processed in the log. That's why I decided to install BattyBirdnet-PI on the Raspberry-Pi 4b.
Audiomoth is connected to the Raspberry-Pi 4b and is used by Battybirdnet-Pi.

BirdNet-Pi runs very well in the docker. That's why I also use the RTSP stream from Audiomoth. My final FFMPEG command is:
`ffmpeg -nostdin -hide_banner -f alsa -ar 48000 -ac 2 -i default -f rtsp -rtsp_transport tcp -ar 48000 -acodec pcm_s16be rtsp://localhost:8554/birdmic
`

Before that, I had the RTSP stream from a surveillance camera. With Audiomoth I now have 3 to 4 times as many recognitions. I use a very good microphone for two applications. :-)

**Comment by Bongo2** [comment]
I still have a problem.
After a reboot, mediamtx and FFMPEG should restart. I entered this in Crontab:

```
@reboot sleep 20; /opt/mediamtx/mediamtx /opt/mediamtx/mediamtx.yml &
@reboot sleep 30; ffmpeg -nostdin -hide_banner -f alsa -ar 48000 -ac 2 -i default -f rtsp -rtsp_transport tcp -ar 48000 -acodec pcm_s16be rtsp://localhost:8554/birdmic &

```
Mediamtx starts. But FFMPEG does not start or ends. I think someone has already described the problem here.
How did your FFMPEG start automatically?

**Comment by tomtom215** [comment]
Hi everyone, 

I have some tools that I've built and have tested on multiple raspberry pi's and multiple USB mics/ sound cards that I think can help the people in this thread that just want to add a USB mic to an RPI and get an audio RTSP stream up and running that will survive reboots. 

I think this will help simplify:

- adding (especially more than one) USB microphones/ sound cards
- downloading & setting up MediaMTX to create and handle RTSP audio streams
- creation of audio streams based on the USB capture devices available to you
- stable in that it survives reboots

I separated the tools into two projects as I think the usb-audio-mapper stands on it's own, but please take a look and try them out. I have it working on a few RPI's but have not tested beyond that. So far it has been more stable than anything I manually stitched together before. 

The ideal starting install situation would be with a dedicated freshly imaged Raspberry Pi that you have SSH access to and at least one USB microphone attached to it and nothing else running/ installed.

1. Add your USB microphone/ sound card one at a time and run this script each time after adding one https://github.com/tomtom215/usb-audio-mapper
2. Once you have mapped the microphones and rebooted, use the scripts in my other project to get RTSP streams up and running found https://github.com/tomtom215/mediamtx-rtsp-setup

Any feedback, GH stars and contribution is welcome, but please do it via the respective project's issues (in other words - I will not handle support in this thread). I may create a dedicated post to this somewhere else, but as this thread is still somewhat alive and was the main place I looked when getting started, I figured there may still be some eyes on it.

Essentially, I have a "microphone Pi" that I setup with multiple mics and used these tools to generate RTSP streams which I added to a dedicated pi running  BirdNET-Pi which allows me to add more mics in other places which has resulted in way more species. 

PS - Shout out to @alexbelgium as they seem to be active in this thread and also adopting this project elsewhere so a GH alert might save you some time in troubleshooting or testing deployments if you find these tools helpful

**Reply by alexbelgium** [reply]
Very nice repo & scripts ! Indeed this is quite useful. Btw I've also written a script that automatically adapts the microphone gain (might need to be adapted for specific microphones) in order to always be in the dynamic range and increase species recognition : https://github.com/alexbelgium/Birdnet-tools/blob/main/autogain.py ; this might be also interesting for your all-in-one system

**Reply by tomtom215** [reply]
> Very nice repo & scripts ! Indeed this is quite useful. Btw I've also written a script that automatically adapts the microphone gain (might need to be adapted for specific microphones) in order to always be in the dynamic range and increase species recognition : https://github.com/alexbelgium/Birdnet-tools/blob/main/autogain.py ; this might be also interesting for your all-in-one system

Thanks for sharing! That's the one piece I haven't automated so far (but have had to troubleshoot manually) and was even thinking about last night, so I'll see if I can incorporate it into my deployment workflow. Thanks again!

**Reply by tomtom215** [reply]
If you want to go from fresh Pi to RTSP audio streaming in (probably) less than 5 minutes with the tools I listed above, try this:

 **Assuming a fresh Raspberry Pi that you access with SSH and have sudo/root permissions and a single USB microphone/ audio capture card connected (you can add more later, see below)** 

First run -> `sudo rasp-pi-config` -> Go to Advanced then Expand File System and reboot

Next, after reboot, run this:

```
sudo apt update && sudo apt upgrade -y && mkdir mediamtx-rtsp-setup && cd mediamtx-rtsp-setup &&
wget https://raw.githubusercontent.com/tomtom215/usb-audio-mapper/main/usb-audio-mapper.sh &&
wget https://raw.githubusercontent.com/tomtom215/mediamtx-rtsp-setup/refs/heads/main/install_mediamtx.sh &&
wget https://raw.githubusercontent.com/tomtom215/mediamtx-rtsp-setup/refs/heads/main/setup_audio_rtsp.sh &&
wget https://raw.githubusercontent.com/tomtom215/mediamtx-rtsp-setup/refs/heads/main/startmic.sh &&
chmod +x usb-audio-mapper.sh install_mediamtx.sh setup_audio_rtsp.sh startmic.sh && sudo ./install_mediamtx.sh && sudo ./usb-audio-mapper.sh
```

If you have more microphones/sound cards to add after rebooting, simply run the usb-audio-mapper.sh script again and reboot for each additional microphone, added one at a time:

`cd mediamtx-rtsp-setup && sudo ./usb-audio-mapper.sh`

After all of your microphones/ sound cards have been added and mapped, or whenever your setup changes, run this to create RTSP streams automatically for every microphone/ sound card you have attached: 

`cd mediamtx-rtsp-setup && sudo ./setup_audio_rtsp.sh`


_Disclaimer: Read the repos and scripts before blindly running scripts from the internet as root._
_Disclaimer 2: I've tested this all-in-one quick setup once so far on an RPI 4, I plan to test a few setups when I have more time, but I have now tested the scripts individually many times with RPI4 and RPI Zero 2W_

**Reply by nommo-san** [reply]
Hi @tomtom215 - thanks for your work on this! It probably saved me a few weeks of frustration. I am trying to get this running on a Pi Zero W as a POC with a view to adding a Griff for Bats and Birds - it's like an Audiomoth).

I think I am 99% of my way there, but I did get this at the last stage (after running cd mediamtx-rtsp-setup && sudo ./setup_audio_rtsp.sh and FFMPEG being installed at the last step):

```
pi@edge:~/mediamtx-rtsp-setup $ sudo check-audio-rtsp.sh
Audio RTSP Service Status Check
Service Status:
Service is running
Running since: 2025-05-28 10:18:19
● audio-rtsp.service - Audio RTSP Streaming Service
     Loaded: loaded (/etc/systemd/system/audio-rtsp.service; enabled; preset: enabled)
     Active: active (running) since Wed 2025-05-28 10:18:19 BST; 4min 39s ago
       Docs: file:/etc/audio-rtsp/config
   Main PID: 2588 (startmic.sh)
      Tasks: 2 (limit: 386)
        CPU: 32.560s
     CGroup: /system.slice/audio-rtsp.service
             ├─2588 /bin/bash /usr/local/bin/startmic.sh
             └─2889 sleep 5

May 28 10:18:19 edgeog systemd[1]: Started audio-rtsp.service - Audio RTSP Streaming Service.
May 28 10:18:21 edgeog useradd[2639]: new user: name=rtsp, UID=994, GID=29, home=/nonexistent, shell=/usr/sbin/nologin, from=none
May 28 10:18:22 edgeog runuser[2654]: pam_unix(runuser:session): session opened for user rtsp(uid=994) by (uid=0)
May 28 10:18:22 edgeog runuser[2654]: pam_unix(runuser:session): session closed for user rtsp

Running Audio Streams:
No active audio streams found

Checking RTSP Server:
RTSP server is accessible on port 18554

Troubleshooting Tips:
1. Check service logs: journalctl -u audio-rtsp -n 50
2. Check ffmpeg status: which ffmpeg && ffmpeg -version
3. Verify MediaMTX is running: systemctl status mediamtx
4. Check config file: cat /etc/audio-rtsp/config
5. Check MediaMTX configuration: cat /etc/mediamtx/mediamtx.yml | grep rtspAddress
6. Manually restart the service: sudo systemctl restart audio-rtsp

Available Sound Cards:
 0 [ugreen         ]: USB-Audio - CM477-30757
                      Solid State System Co.,Ltd. CM477-30757 at usb-20980000.usb-1, full speed
 1 [vc4hdmi        ]: vc4-hdmi - vc4-hdmi
                      vc4-hdmi

Network Information:
To access streams from other devices, use: 192.168.1.48 instead of localhost

System Resources:
awk: cannot open used (No such file or directory)
Memory Usage:
CPU Load: 0.13, 1.22, 1.54
Disk Space: 88% used on log partition

Recent Log Entries:
Last 10 log entries:
[2025-05-28 10:18:54] [WARNING] Skipping card 0 [ugreen] - failed capture test
[2025-05-28 10:18:54] [INFO] Found sound card 1: vc4hdmi - vc4-hdmi - vc4-hdmi
[2025-05-28 10:18:54] [INFO] Skipping card 1 [vc4hdmi] - no capture device found
[2025-05-28 10:18:54] [WARNING] No running RTSP streams detected
[2025-05-28 10:18:54] [WARNING] No audio streams were created. Check if you have audio capture devices connected.
[2025-05-28 10:18:54] [INFO] Running as a systemd service, monitoring streams
[2025-05-28 10:18:54] [INFO] Starting monitor loop for child processes...
[2025-05-28 10:18:54] [INFO] Using RTSP port: 18554
[2025-05-28 10:18:54] [INFO] Service status: Running for 0h 0m, memory usage: 24%
[2025-05-28 10:18:55] [INFO] Active ffmpeg RTSP processes: 00

Error Log:
Error log is empty. No errors reported.

Disk Space for Logs:
8.0K    /var/log/audio-rtsp/
```

It seems my Ugreen USB sound card is failing the capture test and so no stream. Have you seen this behaviour before?

**Reply by nommo-san** [reply]
I tried a few things today to no avail (not surprising really, I do not write code!) so gave up that and am now trying your Github repos. Making more progress.... (I have the ugreen sound card recognised!) this is an FYI more than a support question as I managed to manually install MediaMTX with:

`sudo ./install_mediamtx.sh`

Here's the error I got while running the All-In-One installer - some issue with -v option?

```
[INFO] Successfully downloaded install_mediamtx.sh to /home/pi/install_mediamtx.sh
[INFO] Executing install_mediamtx.sh...
[ERROR] Unknown option: -v
Use --help for usage information
[ERROR] install_mediamtx.sh failed with exit code 1
[INFO] Pre-downloading dependencies for audio RTSP setup...
```

So after manually installing MediaMTX, I've confirmed that my soundcard has a friendly name (ugreen) etc as directed but I still get the error about skipping due to failing capture test:
```

Last 10 log entries:
[2025-05-28 15:29:27] [WARNING] Skipping card 0 [ugreen] - failed capture test
[2025-05-28 15:29:27] [INFO] Found sound card 1: vc4hdmi - vc4-hdmi - vc4-hdmi
[2025-05-28 15:29:27] [INFO] Skipping card 1 [vc4hdmi] - no capture device found
[2025-05-28 15:29:27] [WARNING] No running RTSP streams detected
[2025-05-28 15:29:27] [WARNING] No audio streams were created. Check if you have audio capture devices connected.
[2025-05-28 15:29:27] [INFO] Running as a systemd service, monitoring streams
[2025-05-28 15:29:27] [INFO] Starting monitor loop for child processes...
[2025-05-28 15:29:27] [INFO] Using RTSP port: 18554
[2025-05-28 15:29:27] [INFO] Service status: Running for 0h 0m, memory usage: 23%
[2025-05-28 15:29:28] [INFO] Active ffmpeg RTSP processes: 00

```

**Reply by nommo-san** [reply]
Just a follow up - @tomtom215 did some more work on this and after running this: https://github.com/tomtom215/mediamtx-rtsp-setup I now have a Pi Zero streaming from Ugreen USB soundcard.
